bid-distribution/src/settlement_claims.rs

```diff
@@ -0,0 +1,168 @@
+use crate::sam_meta::ValidatorSamMeta;
+use crate::settlement_config::SettlementConfig;
+use log::info;
+use rust_decimal::prelude::*;
+use rust_decimal::Decimal;
+use settlement_engine::settlement_claims::{
+    Settlement, SettlementClaim, SettlementCollection, SettlementReason,
+};
+use settlement_engine::stake_meta_index::StakeMetaIndex;
+use solana_sdk::pubkey::Pubkey;
+use std::cmp::min;
+use std::collections::HashMap;
+
+pub fn generate_bid_settlement_collection(
+    stake_meta_index: &StakeMetaIndex,
+    sam_validator_metas: &Vec<ValidatorSamMeta>,
+    stake_authority_filter: &dyn Fn(&Pubkey) -> bool,
+    settlement_config: &SettlementConfig,
+) -> SettlementCollection {
+    let settlements = generate_bid_settlements(
+        stake_meta_index,
+        sam_validator_metas,
+        &stake_authority_filter,
+        settlement_config,
+    );
+
+    SettlementCollection {
+        slot: stake_meta_index.stake_meta_collection.slot,
+        epoch: stake_meta_index.stake_meta_collection.epoch,
+        settlements,
+    }
+}
+
+pub fn generate_bid_settlements(
+    stake_meta_index: &StakeMetaIndex,
+    sam_validator_metas: &Vec<ValidatorSamMeta>,
+    stake_authority_filter: &dyn Fn(&Pubkey) -> bool,
+    settlement_config: &SettlementConfig,
+) -> Vec<Settlement> {
+    info!("Generating bid settlements...");
+    let sam_validators_epoch = sam_validator_metas.first().unwrap().epoch;
+
+    assert_eq!(
+        stake_meta_index.stake_meta_collection.epoch, sam_validators_epoch as u64,
+        "SAM Validators Collection epoch must be same as stake meta collection epoch"
+    );
+
+    let marinade_fee_deposit_stake_accounts: HashMap<_, _> = stake_meta_index
+        .stake_meta_collection
+        .stake_metas
+        .iter()
+        .find(|x| {
+            x.withdraw_authority
+                .eq(settlement_config.marinade_withdraw_authority())
+                && x.stake_authority
+                    .eq(settlement_config.marinade_stake_authority())
+        })
+        .iter()
+        .map(|s| (s.pubkey, s.active_delegation_lamports))
+        .collect();
+
+    let mut settlement_claim_collections = vec![];
+
+    for validator in sam_validator_metas {
+        if let Some(grouped_stake_metas) =
+            stake_meta_index.iter_grouped_stake_metas(&validator.vote_account)
+        {
+            if validator.effective_bid == Decimal::ZERO {
```

Comment:
I like this. I should fix the protected events processing in the same way (what I haven't done despite I check with you here, blame on me :-))